src/self_driving/planning/seq_tuning/graph_solver.cpp

```diff
@@ -0,0 +1,146 @@
+#include "self_driving/planning/seq_tuning/graph_solver.h"
+
+#include <cmath>
+
+#include "loggers/selfdriving_logger.h"
+#include "self_driving/forecasting/workload_forecast.h"
+#include "self_driving/planning/action/abstract_action.h"
+#include "self_driving/planning/pilot.h"
+#include "self_driving/planning/pilot_util.h"
+#include "self_driving/planning/seq_tuning/path_solution.h"
+
+namespace noisepage::selfdriving::pilot {
+
+GraphSolver::GraphSolver(const PlanningContext &planning_context,
+                         common::ManagedPointer<selfdriving::WorkloadForecast> forecast, uint64_t end_segment_index,
+                         const std::map<action_id_t, std::unique_ptr<AbstractAction>> &structure_map,
+                         const std::vector<double> &default_segment_cost,
+                         const std::vector<std::set<std::set<action_id_t>>> &candidate_configurations_by_segment,
+                         uint64_t memory_constraint) {
+  // build the graph and compute shortest path simultaneously
+
+  ActionState action_state;
+  action_state.SetIntervals(0, end_segment_index);
+
+  source_level_.push_back(std::make_unique<SeqNode>(std::set<action_id_t>(), 0.0, 0.0, action_state, true));
+
+  auto new_action_state = action_state;
```

Comment:
It seems that `action_state` is never used except in the final "dummy destination node". So the `new_action_state` seems unnecessary?